File: solutions/permutations/permutations.cpp
// @Title: Permutations
// @Author: colinjxc
// @Date: 2022-03-05T13:55:00+08:00
// @URL: https://leetcode-cn.com/problems/permutations


class Solution {
public:
    vector<vector<int>> res;
    vector<bool> vis;
    vector<vector<int>> permute(vector<int>& nums) {
        res.clear();
        vis.resize(nums.size());

        vector<int> tmp;
        backtrack(nums, tmp, 0);
        return res;
    }

    void backtrack(vector<int>& nums, vector<int>& tmp, int index) {
        if (index == nums.size()) {
            res.push_back(tmp);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (!vis[i]) {
                vis[i] = true;
                tmp.push_back(nums[i]);
                backtrack(nums, tmp, index+1);
                tmp.pop_back();
                vis[i] = false;
            }
        }
    }
};

File: solutions/permutations/permutations.py
// @Title: Permutations
// @Author: colinjxc
// @Date: 2022-03-19T04:59:54+08:00
// @URL: https://leetcode-cn.com/problems/permutations


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        def backtrack(nums, tmp):
            if not nums:
                res.append(tmp)
                return 
            for i in range(len(nums)):
                backtrack(nums[:i] + nums[i+1:], tmp + [nums[i]])
                
        backtrack(nums, [])
        return res

